pkg/client/apiutil/restmapper_test.go

```diff
@@ -564,6 +550,131 @@ func TestLazyRestMapperProvider(t *testing.T) {
+		err = apiextensionsv1.AddToScheme(s)
+		g.Expect(err).NotTo(gmg.HaveOccurred())
+
+		c, err := client.New(restCfg, client.Options{Scheme: s})
+		g.Expect(err).NotTo(gmg.HaveOccurred())
+
+		// Register a new CRD ina  new group to avoid collisions when deleting versions - "taxi.inventory.example.com".
+		group := "inventory.example.com"
+		kind := "Taxi"
+		plural := "taxis"
+		crdName := plural + "." + group
+		crd := createNewCRD(ctx, g, c, group, kind, plural)
+		t.Cleanup(func() {
+			g.Expect(c.Delete(ctx, crd)).To(gmg.Succeed())
+		})
+
+		// Wait until the CRD is registered.
+		discHTTP, err := rest.HTTPClientFor(restCfg)
+		g.Expect(err).NotTo(gmg.HaveOccurred())
+		discClient, err := discovery.NewDiscoveryClientForConfigAndClient(restCfg, discHTTP)
+		g.Expect(err).NotTo(gmg.HaveOccurred())
+		g.Eventually(func(g gmg.Gomega) {
+			_, err = discClient.ServerResourcesForGroupVersion(group + "/v1")
+			g.Expect(err).NotTo(gmg.HaveOccurred())
+		}).Should(gmg.Succeed(), "v1 should be available")
+
+		// There are no requests before any call
+		g.Expect(crt.GetRequestCount()).To(gmg.Equal(0))
+
+		// Since we don't specify what version we expect, restmapper will fetch them all and search there.
+		// To fetch a list of available versions
+		//  #1: GET https://host/api
+		//  #2: GET https://host/apis
+		// Then, for all available versions:
+		// 	#3: GET https://host/apis/inventory.example.com/v1
+		//	#4: GET https://host/apis/inventory.example.com/v2
+		// This should fill the cache for apiGroups and versions.
+		mapping, err := lazyRestMapper.RESTMapping(schema.GroupKind{Group: group, Kind: kind})
+		g.Expect(err).NotTo(gmg.HaveOccurred())
+		g.Expect(mapping.GroupVersionKind.Kind).To(gmg.Equal(kind))
+		g.Expect(crt.GetRequestCount()).To(gmg.Equal(4))
+		crt.Reset() // We reset the counter to check how many additional requests are made later.
+
+		// At this point v2 should be cached
+		_, err = lazyRestMapper.RESTMapping(schema.GroupKind{Group: group, Kind: kind}, "v2")
+		g.Expect(err).NotTo(gmg.HaveOccurred())
+		g.Expect(crt.GetRequestCount()).To(gmg.Equal(0))
+
+		// We update the CRD to only have v1 version.
+		g.Expect(c.Get(ctx, types.NamespacedName{Name: crdName}, crd)).To(gmg.Succeed())
+		var v1 apiextensionsv1.CustomResourceDefinitionVersion
+		for i, version := range crd.Spec.Versions {
+			if version.Name == "v1" {
+				crd.Spec.Versions[i].Storage = true
+				v1 = version
+				v1.Storage = true
+			}
+		}
+		crd.Spec.Versions = []apiextensionsv1.CustomResourceDefinitionVersion{v1}
+		g.Expect(c.Update(ctx, crd)).To(gmg.Succeed())
+
+		// We wait until v2 is not available anymore.
+		g.Eventually(func(g gmg.Gomega) {
+			_, err = discClient.ServerResourcesForGroupVersion(group + "/v2")
+			g.Expect(apierrors.IsNotFound(err)).To(gmg.BeTrue(), "v2 should not be available anymore")
+		}).Should(gmg.Succeed())
+
+		// Although v2 is not available anymore, the cache is not invalidated yet so it should return a mapping.
```

Comment:
Just a question. I didn't realize this before, but once the CAPI controllers are all running for a bit, they wouldn't hit the meta.IsNoMatchError case, right?

So this only happens if the apiVersion goes away at the wrong time and then (in CAPI) we do an external.Get on a resource that isn't already cached?

Reply:
Yeah. From capi's perspective, the error happens (only in v0.16.x) when a group is cached in m.apiGroups with a particular version that will be deleted. If after this is cached, we request a resource that for some reason is not yet in the underlying RESTMapper, that triggers a "re-read" of all the cached versions for the resources's group. If one of these cached versions have been deleted, it will return an error. This will keep happening forever since m.apigroups is not cleaned.

In v0.17.x this "re-read" for a non-existing version still happens every time there is a miss in the underlying RESTMapper, but the error is ignored, so the caller continues as normal.

Reply:
Perfect, thx!